File: 543-diameter-of-binary-tree/diameter-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private int maxDiameter(TreeNode root, int[] arr){
        if(root == null) return 0;
        int left = maxDiameter(root.left, arr);
        int right = maxDiameter(root.right, arr);

        arr[0] = Math.max(left+right,arr[0]);

        return 1 + Math.max(left,right);
    }
    public int diameterOfBinaryTree(TreeNode root) {
        int[] arr = {0};
        maxDiameter(root,arr);
        return arr[0];
    }
}

File: 543-diameter-of-binary-tree/diameter-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        self.maxi = 0
        def depth(node):
            if not node: return 0
            lh = depth(node.left)
            rh = depth(node.right)
            self.maxi = max(self.maxi,lh+rh)
            return 1 + max(lh,rh)
        self.maxi
        depth(root)
        return self.maxi